File: collectors/code-analyzer/src/publisher.py
"""CloudEvents publisher for code analysis results."""

import json
import logging
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

import aio_pika

logger = logging.getLogger(__name__)


class EventPublisher:
    """Publishes CloudEvents to RabbitMQ."""

    def __init__(
        self,
        channel: aio_pika.Channel,
        exchange_name: str,
        scan_id: str | None = None,  # ADR-007: scan_id for CloudEvent subject
    ):
        self.channel = channel
        self.exchange_name = exchange_name
        self._exchange: aio_pika.Exchange | None = None
        self._scan_id = scan_id  # ADR-007: Used as CloudEvent subject

    def set_scan_id(self, scan_id: str | None) -> None:
        """Set the scan_id for CloudEvent subject (ADR-007)."""
        self._scan_id = scan_id

    async def _get_exchange(self) -> aio_pika.Exchange:
        """Get or create the exchange."""
        if self._exchange is None:
            self._exchange = await self.channel.declare_exchange(
                self.exchange_name,
                aio_pika.ExchangeType.TOPIC,
                durable=True,
            )
        return self._exchange

    def _create_cloudevent(
        self,
        event_type: str,
        data: dict[str, Any],
        source: str = "code-analyzer",
    ) -> dict[str, Any]:
        """Create a CloudEvent envelope."""
        event = {
            "specversion": "1.0",
            "id": str(uuid4()),
            "source": f"/collectors/{source}",  # ADR-007: Standardized source
            "type": event_type,
            "time": datetime.now(timezone.utc).isoformat(),
            "datacontenttype": "application/json",
            "data": data,
        }

        # ADR-007: Set subject = scan_id for orchestration tracking
        if self._scan_id:
            event["subject"] = self._scan_id

        return event

    async def _publish(
        self,
        routing_key: str,
        event_type: str,
        data: dict[str, Any],
    ) -> None:
        """Publish a CloudEvent."""
        exchange = await self._get_exchange()
        event = self._create_cloudevent(event_type, data)

        message = aio_pika.Message(
            body=json.dumps(event).encode(),
            content_type="application/cloudevents+json",
            delivery_mode=aio_pika.DeliveryMode.PERSISTENT,
        )

        await exchange.publish(message, routing_key=routing_key)
        logger.info(f"Published {event_type} event to {routing_key}")

    async def publish_repository_discovered(
        self,
        analysis_id: str,
        repo_url: str,
        branch: str,
        languages: dict[str, Any],
        frameworks: list[dict[str, Any]],
    ) -> None:
        """
        Publish repository discovered event.

        CloudEvents type: discovery.repository.discovered
        Routing key: discovered.repository
        """
        data = {
            "analysis_id": analysis_id,
            "repository_url": repo_url,
            "branch": branch,
            "languages": languages,
            "frameworks": frameworks,
            "discovered_at": datetime.now(timezone.utc).isoformat(),
        }

        await self._publish(
            routing_key="discovered.repository",
            event_type="discovery.repository.discovered",
            data=data,
        )

    async def publish_codebase_discovered(
        self,
        analysis_id: str,
        repo_url: str,
        metrics: dict[str, Any],
    ) -> None:
        """
        Publish codebase metrics discovered event.

        CloudEvents type: discovery.codebase.discovered
        Routing key: discovered.codebase
        """
        data = {
            "analysis_id": analysis_id,
            "repository_url": repo_url,
            "metrics": metrics,
            "discovered_at": datetime.now(timezone.utc).isoformat(),
        }

        await self._publish(
            routing_key="discovered.codebase",
            event_type="discovery.codebase.discovered",
            data=data,
        )

    async def publish_dependency_discovered(
        self,
        analysis_id: str,
        repo_url: str,
        dependency: dict[str, Any],
    ) -> None:
        """
        Publish dependency discovered event.

        CloudEvents type: discovery.dependency.discovered
        Routing key: discovered.dependency
        """
        data = {
            "analysis_id": analysis_id,
            "repository_url": repo_url,
            "dependency": dependency,
            "discovered_at": datetime.now(timezone.utc).isoformat(),
        }

        await self._publish(
            routing_key="discovered.dependency",
            event_type="discovery.dependency.discovered",
            data=data,
        )

File: collectors/network-scanner/internal/publisher/publisher.go
// Package publisher handles publishing discovery events to RabbitMQ.
package publisher

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	amqp "github.com/rabbitmq/amqp091-go"
	"go.uber.org/zap"
)

// Publisher sends CloudEvents to RabbitMQ.
type Publisher struct {
	conn     *amqp.Connection
	channel  *amqp.Channel
	exchange string
	logger   *zap.SugaredLogger
	scanID   string // ADR-007: Current scan ID for CloudEvent subject
}

// CloudEvent represents the CloudEvents 1.0 specification structure.
type CloudEvent struct {
	SpecVersion     string      `json:"specversion"`
	Type            string      `json:"type"`
	Source          string      `json:"source"`
	ID              string      `json:"id"`
	Subject         string      `json:"subject,omitempty"` // scan_id for orchestration (ADR-007)
	Time            string      `json:"time"`
	DataContentType string      `json:"datacontenttype"`
	Data            interface{} `json:"data"`
}

// ServerDiscoveredData represents data for a discovered server event.
type ServerDiscoveredData struct {
	ServerID    string   `json:"server_id"`
	Hostname    string   `json:"hostname,omitempty"`
	IPAddresses []string `json:"ip_addresses"`
	OpenPorts   []int    `json:"open_ports"`
	OS          *OSInfo  `json:"os,omitempty"`
}

// ServiceDiscoveredData represents data for a discovered service event.
type ServiceDiscoveredData struct {
	ServiceID string                 `json:"service_id"`
	ServerID  string                 `json:"server_id"`
	IP        string                 `json:"ip"`
	Port      int                    `json:"port"`
	Protocol  string                 `json:"protocol"`
	Service   string                 `json:"service,omitempty"`
	Version   string                 `json:"version,omitempty"`
	Banner    string                 `json:"banner,omitempty"`
	Metadata  map[string]interface{} `json:"metadata,omitempty"` // ADR-007: candidate flags
}

// Database ports for candidate identification (ADR-007)
var databasePorts = map[int]string{
	3306:  "mysql",
	5432:  "postgresql",
	27017: "mongodb",
	6379:  "redis",
	1433:  "mssql",
	1521:  "oracle",
	9200:  "elasticsearch",
	5984:  "couchdb",
}

// OSInfo contains operating system information.
type OSInfo struct {
	Name    string `json:"name,omitempty"`
	Version string `json:"version,omitempty"`
	Family  string `json:"family,omitempty"`
}

// New creates a new Publisher connected to RabbitMQ.
func New(url string, logger *zap.SugaredLogger) (*Publisher, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RabbitMQ: %w", err)
	}

	channel, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("failed to open channel: %w", err)
	}

	return &Publisher{
		conn:     conn,
		channel:  channel,
		exchange: "discovery.events",
		logger:   logger,
	}, nil
}

// Close closes the RabbitMQ connection.
func (p *Publisher) Close() error {
	if p.channel != nil {
		_ = p.channel.Close()
	}
	if p.conn != nil {
		return p.conn.Close()
	}
	return nil
}

// SetScanID sets the current scan ID for CloudEvent subject (ADR-007).
func (p *Publisher) SetScanID(scanID string) {
	p.scanID = scanID
}

// GetScanID returns the current scan ID.
func (p *Publisher) GetScanID() string {
	return p.scanID
}

// PublishServerDiscovered publishes a server discovered event.
func (p *Publisher) PublishServerDiscovered(data ServerDiscoveredData) error {
	event := p.createEvent("discovery.server.discovered", data)
	return p.publish(event, "discovered.server")
}

// PublishServiceDiscovered publishes a service discovered event.
func (p *Publisher) PublishServiceDiscovered(result interface{}) error {
	// Convert ScanResult to ServiceDiscoveredData
	var data ServiceDiscoveredData

	// Type assertion for scanner.ScanResult
	if scanResult, ok := result.(interface {
		GetIP() string
		GetPort() int
		GetProtocol() string
		GetService() string
		GetBanner() string
	}); ok {
		port := scanResult.GetPort()
		banner := scanResult.GetBanner()

		data = ServiceDiscoveredData{
			ServiceID: uuid.New().String(),
			IP:        scanResult.GetIP(),
			Port:      port,
			Protocol:  scanResult.GetProtocol(),
			Service:   scanResult.GetService(),
			Banner:    banner,
			Metadata:  buildMetadata(port, banner), // ADR-007: Add candidate flags
		}
	} else {
		// Direct struct conversion for simple cases
		jsonBytes, err := json.Marshal(result)
		if err != nil {
			return fmt.Errorf("failed to marshal result: %w", err)
		}
		if err := json.Unmarshal(jsonBytes, &data); err != nil {
			return fmt.Errorf("failed to unmarshal to ServiceDiscoveredData: %w", err)
		}
		if data.ServiceID == "" {
			data.ServiceID = uuid.New().String()
		}
		// Add metadata if not present
		if data.Metadata == nil {
			data.Metadata = buildMetadata(data.Port, data.Banner)
		}
	}

	event := p.createEvent("discovery.service.discovered", data)
	return p.publish(event, "discovered.service")
}

func (p *Publisher) createEvent(eventType string, data interface{}) CloudEvent {
	event := CloudEvent{
		SpecVersion:     "1.0",
		Type:            eventType,
		Source:          "/collectors/network-scanner",
		ID:              uuid.New().String(),
		Time:            time.Now().UTC().Format(time.RFC3339),
		DataContentType: "application/json",
		Data:            data,
	}

	// Set subject = scan_id for orchestration tracking (ADR-007)
	if p.scanID != "" {
		event.Subject = p.scanID
	}

	return event
}

// buildMetadata creates metadata with database candidate flags (ADR-007).
func buildMetadata(port int, banner string) map[string]interface{} {
	metadata := make(map[string]interface{})

	// Check if port is a known database port
	if dbType, ok := databasePorts[port]; ok {
		metadata["database_candidate"] = true
		metadata["candidate_type"] = dbType
		metadata["candidate_confidence"] = 0.5 // port_only confidence
		metadata["candidate_reason"] = fmt.Sprintf("Port %d (known %s port)", port, dbType)
	}

	return metadata
}

func (p *Publisher) publish(event CloudEvent, routingKey string) error {
	body, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	err = p.channel.PublishWithContext(
		ctx,
		p.exchange,
		routingKey,
		false, // mandatory
		false, // immediate
		amqp.Publishing{
			ContentType: "application/cloudevents+json",
			Body:        body,
			MessageId:   event.ID,
			Timestamp:   time.Now(),
		},
	)

	if err != nil {
		return fmt.Errorf("failed to publish event: %w", err)
	}

	p.logger.Debugw("Event published",
		"type", event.Type,
		"id", event.ID,
		"routing_key", routingKey,
	)

	return nil
}
